### PR TITLE
[types] Add type definitions from DefinitelyTyped/DefinitelyTyped#62510

### LINKED ISSUE #119: Typescript support
Are there any plans to support typescript when version 1.0 releases?

### PR DESCRIPTION
**Pull request by MysteryBlokHed**
Closes #119

Moves types from the original DefinitelyTyped PR and adds some dependencies and workflows to test them.

This PR is still very much a work in progress, and I'm pretty sure there are some parts of the API that I haven't typed yet.

### CODE REVIEW

**Comment by LeaVerou** [comment]
Wow, that is a very substantial amount of work, thank you so much @MysteryBlokHed!

I'm a bit worried I lack the TS experience to review this properly, perhaps someone else could take a look when it's ready? @ai @jgerigmeyer would that be of interest to you?

**Comment by jgerigmeyer** [comment]
> @jgerigmeyer would that be of interest to you?

Yes, I'd be happy to take a look at this! I should be able to do that in the next few days. 👍

**Review by jgerigmeyer — CHANGES_REQUESTED** [review]
This is so helpful -- thanks @MysteryBlokHed! 🎉 

I wasn't able to get dtslint running locally without some further changes -- submitted in https://github.com/MysteryBlokHed/color.js/pull/1

I pulled this in to a personal TypeScript project that uses colorjs.io (replacing some type stubs I'd written there), and there were only a few places where it seems these types were slightly incorrect -- I submitted a first pass at some changes in https://github.com/MysteryBlokHed/color.js/pull/2

**Review by jgerigmeyer — APPROVED** [review]
🚀

**Comment by LeaVerou** [comment]
@jgerigmeyer thank you so much! I just added you as a collaborator. 

@MysteryBlokHed @jgerigmeyer is this ready to merge?

File: .github/workflows/test-types.yml
Comment (LeaVerou) [inline .github/workflows/test-types.yml:12]: Isn't 12 a bit old at this point?

File: .gitignore
Comment (LeaVerou) [inline .gitignore:1]: Why?

File: package.json
Comment (MysteryBlokHed) [reply package.json:17]: I just kept this structure when I copied the types over from DefinitelyTyped, but I think doing it this way also makes testing easier since everything's in the same place

**Comment by LeaVerou** [comment]
I had a quick look and left some questions.

I see there is also a new `test` directory with different tests. How does this compare to the existing tests? Are they complementary? Is it primarily testing types? We should probably document why there are two different sets of tests somewhere, for new contributors?

File: .gitignore
Comment (MysteryBlokHed) [reply .gitignore:1]: Adding the leading slash means that Git will only ignore the `dist` folder in the root of the project. Without this change, the folder `types/dist` is also ignored

File: .github/workflows/test-types.yml
Comment (MysteryBlokHed) [reply .github/workflows/test-types.yml:12]: I think I chose 12 because the Babel config specifies version <=12, but it could be made more re ent

**Comment by jgerigmeyer** [comment]
I'm 99% sure that we'll discover small issues once people start actually using this with TypeScript, but IMO this is already a significant improvement and is ready to merge. 🎉 

It would be much easier to maintain these types going forward if the codebase itself were re-written in TypeScript, but I understand that's not a simple change and wouldn't have to happen right away regardless.

**Comment by MysteryBlokHed** [comment]
The tests in the `types` directory only test the definitions; they're not actually evaluated like the existing tests.

A mention of the reason for the two directories could be mentioned in the README or CONTRIBUTING files maybe? I could make a change when I'm free in a few hours

**Comment by LeaVerou** [comment]
> > I see there is also a new test directory with different tests. How does this compare to the existing tests? Are they complementary? Is it primarily testing types? We should probably document why there are two different sets of tests somewhere, for new contributors?
> 
> The tests in the `types` directory only test the definitions; they're not actually evaluated like the existing tests.
> 
> A mention of the reason for the two directories could be mentioned in the README or CONTRIBUTING files maybe? I could make a change when I'm free in a few hours

I totally missed they were in `types`. I think that's pretty self-explanatory, though of course you're welcome to add something in CONTRIBUTING.md if you feel like it.

**Comment by jgerigmeyer** [comment]
> Ok, I plan to rebase and merge soon, so if there's anything you want to squash in the commit log, now's the time.

I'm not particular about a "clean" git history, so I'll defer to @MysteryBlokHed on whether to squash before merging.

**Comment by MysteryBlokHed** [comment]
Rebasing seems fine and I don't think there's anything else that needs changing right now

**Comment by LeaVerou** [comment]
Merged, thank you both!

**Comment by 1nfinity-5starZ** [comment]
Was this shipped in latest release ?

**Comment by LeaVerou** [comment]
No, it will be in the next release (likely v0.4.1)

**Pull request merged**
